irods/exception.py

```diff
@@ -67,8 +69,39 @@ def __init__(self, name, bases, attrs):
             iRODSExceptionMeta.codes[attrs['code']] = self
 
 
+class Errno:
+    def __init__(self,arg0,*args):
+        self.int_code = arg0
+    def __repr__(self):
+        e = self.int_code
+        try:
+            return self.__class__.__name__ + repr(tuple([e, errno.errorcode[e], os.strerror(e)]))
+        except Exception as exc:
+            pass # print(repr(exc))
+        return self.__class__.__name__ + repr(tuple([e,]))
+    def __int__(self):
+        return self.int_code
+
+
 class iRODSException(six.with_metaclass(iRODSExceptionMeta, Exception)):
-    pass
+
+    def __init__(self,*arg):
+        explicit_errno = None
+        if hasattr(self.__class__,'code'):
+           argl = list(arg)
+           if argl and isinstance (argl[-1],Errno):
+               explicit_errno = argl.pop()
+        super(iRODSException,self).__init__(*argl)
+        if explicit_errno:
+            self.code -= int(explicit_errno)
```

Comment:
It could be helpful to add some comments explaining why we are doing this. This has to do with the OS error codes being encoded into the iRODS error codes (sometimes), right?